tests/patch_inline_callbacks.py

```diff
@@ -92,3 +101,79 @@ def check_ctx(r):
         return wrapped
 
     defer.inlineCallbacks = new_inline_callbacks
+    new_inline_callbacks.patched_by_synapse = True
+
+
+def _check_yield_points(f, changes, start_context):
+    """Wraps a generator that is about to passed to defer.inlineCallbacks
+    checking that after every yield the log contexts are correct.
+    """
+
+    from synapse.logging.context import LoggingContext
+
+    @functools.wraps(f)
+    def check_yield_points_inner(*args, **kwargs):
+        expected_context = start_context
+
+        gen = f(*args, **kwargs)
+
+        last_yield_line_no = 1
+        result = None
+        while True:
+            try:
```

Comment:
I'd be inclined just to stick an `expected_context = LoggingContext.current_context()` here instead of line 116 and 175, but ymmv